README.md
# nsu_labs
## Usage

    $ lein run


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-12-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-12-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/nsu_labs/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/nsu_labs/compare/0.1.0...0.1.1


test/nsu_labs/core_test.clj
(ns nsu-labs.core-test
    (:require [clojure.test :refer :all]
              [nsu-labs.core :refer :all]))


(deftest filter-tests
  (is
   (=
    (filter my-wierd-predicate (range 1000000))
    (parallel-filter my-wierd-predicate (range 1000000))))
  (is
   (=
    (parallel-filter (fn [junk] true) (range 1000000))
    (range 1000000)))
  (is
   (=
    (parallel-filter (fn [junk] false) (range 1000000))
    '()))
  (is
   (= (take 9 (filter prime? (iterate inc 1)))
      (take 9 (parallel-filter prime? (iterate inc 1))))))

src_copy.clj
(ns nsu-labs.core
    (:gen-class))

(defn parallel-filter
  ([predicate coll]
   (let [parts       (map #(partition-all 2048 %) (take 16384 coll))
         black-magic (fn [coll2]
                       (doall
                        (reduce concat '()
                                (doall
                                 (map deref
                                      (doall (map (fn [coll1] (future (doall (filter predicate coll1)))) coll2)))))))]
     (lazy-seq (concat (reduce concat (map black-magic parts)) (drop 16384 coll))))))

(defn prime? [n]
  (if (even? n)
    false
    (let [root (num (int (Math/sqrt n)))]
      (loop [i 3]
        (if (> i root)
          true
          (if (zero? (mod n i))
            false
            (recur (+ i 2))))))))

(defn my-wierd-predicate [x]
  (and (prime? x) (= (mod x 1337) 22)))
(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Usual filter:")
  (println (time (doall (filter my-wierd-predicate (range 1000000)))))
  (println (time (doall (filter my-wierd-predicate (range 1000000)))))
  (println (time (doall (filter my-wierd-predicate (range 1000000)))))
  (println "My parallel filter:")
  (println (time (doall (parallel-filter my-wierd-predicate (range 1000000)))))
  (println (time (doall (parallel-filter my-wierd-predicate (range 1000000)))))
  (println (time (doall (parallel-filter my-wierd-predicate (range 1000000)))))
  (println "lazyness test")
  (println (take 1 (parallel-filter my-wierd-predicate (iterate inc 1))))
  (shutdown-agents))



src/nsu_labs/agents.clj
(ns nsu-labs.agents
    (:gen-class))

(declare supply-msg)
(declare notify-msg)

(let [lock (Object.)]
  (defn sync-println [& args]
    (locking lock (apply println args))))


(defn storage
  "Creates a new storage
   ware - a name of ware to store (string)
   notify-step - amount of stored items required for logger to react. 0 means to logging
   consumers - factories to notify when the storage is updated
   returns a map that contains:
     :storage - an atom to store items that can be used by factories directly
     :ware - a stored ware name
     :worker - an agent to send supply-msg"
  [ware notify-step & consumers]
  (let [counter      (atom 0 :validator #(>= % 0)),
        worker-state {:storage     counter,
                      :ware        ware,
                      :notify-step notify-step,
                      :consumers   consumers}]
    {:storage counter,
     :ware    ware,
     :worker  (agent worker-state)}))

(defn factory
  "Creates a new factory
   amount - number of items produced per cycle
   duration - cycle duration in milliseconds
   target-storage - a storage to put products with supply-msg
   ware-amounts - a list of ware names and their amounts required for a single cycle
   returns a map that contains:
     :worker - an agent to send notify-msg"
  [amount duration target-storage & ware-amounts]
  (let [bill         (apply hash-map ware-amounts),
        buffer       (reduce-kv (fn [acc k _] (assoc acc k 0))
                                {} bill),
        ;;a state of factory agent:
        ;;  :amount - a number of items to produce per cycle
        ;;  :duration - a duration of cylce
        ;;  :target-storage - a storage to place products (via supply-msg to its worker)
        ;;  :bill - a map with ware names as keys and their amounts of values
        ;;     shows how many wares must be consumed to perform one production cycle
        ;;  :buffer - a map with similar structure as for :bill that shows how many wares are already collected;
        ;;     it is the only mutable part.
        worker-state {:amount         amount,
                      :duration       duration,
                      :target-storage target-storage,
                      :bill           bill,
                      :buffer         buffer}]
    {:worker (agent worker-state)}))

(defn source
  "Creates a source that is a thread that produces 'amount' of wares per cycle to store in 'target-storage'
   and with given cycle 'duration' in milliseconds
   returns Thread that must be run explicitly"
  [amount duration target-storage]
  (new Thread
    (fn []
      (Thread/sleep duration)
      (send (target-storage :worker) supply-msg amount)
      (recur))))

(defn supply-msg
  "A message that can be sent to a storage worker to notify that the given 'amount' of wares should be added.
   Adds the given 'amount' of ware to the storage and notifies all the registered factories about it
   state - see code of 'storage' for structure"
  [state amount]
  (swap! (state :storage) #(+ % amount)) ;update counter, could not fail
  (let [ware        (state :ware),
        cnt         @(state :storage),
        notify-step (state :notify-step),
        consumers   (state :consumers)]
    ;;logging part, notify-step == 0 means no logging
    (when
     (and (> notify-step 0)
          (> (int (/ cnt notify-step))
             (int (/ (- cnt amount) notify-step))))
     (sync-println
      (.format (new java.text.SimpleDateFormat "hh.mm.ss.SSS") (new java.util.Date))
      "|" ware "amount: " cnt))
    ;;factories notification part
    (when consumers
          (doseq [consumer (shuffle consumers)]
            (send (consumer :worker) notify-msg ware (state :storage) amount))))
  state)

;worker itself is immutable, keeping configuration only


(defn notify-msg
  "A message that can be sent to a factory worker to notify that the provided 'amount' of 'ware's are
   just put to the 'storage-atom'."
  ;; 'state' is for agent created in 'factory', see comments in its code for details
  ;;The implementation should:
  ;; - try to retrieve some items from the 'storage-atom' if necessary
  ;; - if the retrieval is not successful, do not forget to handle validation exception correctly
  ;; - if the retrieval is successful, put wares into the internal ':buffer'
  ;; - when there are enough wares of all types according to :bill, a new cycle must be started with given duratin;
  ;;   after it finished all the wares must be removed from the internal ':buffer' and ':target-storage' must be notified
  ;;   with 'supply-msg'
  ;; - return new agent state with possibly modified ':buffer' in any case!
  [state ware storage-atom amount]
  ;;;;;;;;;;;;;;;;;;;;;;;
  ;;TODO implement me
  ;;;;;;;;;;;;;;;;;;;;;;;
  (let [required      (fn [w] (get (get state :bill {}) w 0)),
        willTake      (min (required ware) @storage-atom),
        canProduce    (every?
                       (fn [key] (>= (get (get state :bill) key 0) (get (get state :buffer) key 0)))
                       (keys (get state :bill))),
        filledState   (update state :buffer (fn [m] (update m ware + willTake))),
        producedState (fn [oldState]
                        (reduce (fn [acc k] (update acc k - (required k))) oldState (keys (get state :bill))))]
    (if (> willTake 0)
      (try
        ((sync-println "Taking" willTake ware)
          (swap! storage-atom (fn [x] (- x willTake)))
          (sync-println "Before:" (get state :buffer))
          (sync-println "After:" (get filledState :buffer))
          (if canProduce
            ((Thread/sleep (get state :duration))
              (send (get (get state :target-storage) :worker) supply-msg (get state :amount))
              (sync-println "Produced an item!")
              producedState)
            filledState))
        (catch Exception e state))
      state)))


;;;
(def safe-storage (storage "Safe" 1))
(def safe-factory (factory 1 3000 safe-storage "Metal" 3))
(def cuckoo-clock-storage (storage "Cuckoo-clock" 1))
(def cuckoo-clock-factory
  (factory 1 2000 cuckoo-clock-storage "Lumber" 5 "Gears" 10))
(def gears-storage (storage "Gears" 20 cuckoo-clock-factory))
(def gears-factory (factory 4 1000 gears-storage "Ore" 4))
(def metal-storage (storage "Metal" 5 safe-factory))
(def metal-factory (factory 1 1000 metal-storage "Ore" 10))
(def lumber-storage (storage "Lumber" 20 cuckoo-clock-factory))
(def lumber-mill (source 5 4000 lumber-storage))
(def ore-storage (storage "Ore" 10 metal-factory gears-factory))
(def ore-mine (source 2 1000 ore-storage))

;;;runs sources and the whole process as the result
(defn start []
  (.start ore-mine)
  (.start lumber-mill))

;;;stopes running process
;;;recomplile the code after it to reset all the process
(defn stop []
  (.stop ore-mine)
  (.stop lumber-mill))

;;;This could be used to aquire errors from workers
;;;(agent-error (gears-factory :worker))
;;;(agent-error (metal-storage :worker))

(defn -main
  [& args]
  (start)
  (Thread/sleep 60000)
  (stop)
  (Thread/sleep 2000)
  (shutdown-agents))

src/nsu_labs/routes.clj
(ns nsu-labs.routes
    (:gen-class))
;;;an empty route map
;;;it is enough to use either forward or backward part (they correspond to each other including shared reference to number of tickets)
;;;:forward is a map with route start point names as keys and nested map as values
;;;each nested map has route end point names as keys and route descriptor as values
;;;each route descriptor is a map (structure in fact) of the fixed structure where 
;;;:price contains ticket price
;;;and :tickets contains reference to tickets number
;;;:backward has the same structure but start and end points are reverted 
(def empty-map
  {:forward  {},
   :backward {}})

(defn route
  "Add a new route (route) to the given route map
   route-map - route map to modify
   from - name (string) of the start point of the route
   to - name (string) of the end poiunt of the route
   price - ticket price
   tickets-num - number of tickets available"
  [route-map from to price tickets-num]
  (let [tickets                (ref tickets-num :validator (fn [state] (>= state 0))),     ;reference for the number of tickets
        orig-source-desc       (or (get-in route-map [:forward from]) {}),
        orig-reverse-dest-desc (or (get-in route-map [:backward to]) {}),
        route-desc             {:price   price,
                                ;route descriptor
                                :tickets tickets},
        source-desc            (assoc orig-source-desc to route-desc),
        reverse-dest-desc      (assoc orig-reverse-dest-desc from route-desc)]
    (-> route-map
        (assoc-in [:forward from] source-desc)
        (assoc-in [:backward to] reverse-dest-desc))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(def inf 999999)

; default dijksta algorithm
(defn dijkstra [g src dst]
  (loop [dists (assoc (zipmap (keys g) (repeat inf)) src 0)
         curr src ;
         ns (keys (get g curr)) ; available moves
         unvisited (apply hash-set (keys g))
         p (zipmap (keys g) (repeat nil))] ; paths
    (if (empty? unvisited) ; if all cities are visited, shortest path is found
      [(get dists dst),
       (loop [path []
              t dst]
         (if (= t src)
           (cons src path)
           (recur (cons t path) (get p t))))]
      (if (empty? ns)
        (let [u (disj unvisited curr)
              next-n (first (sort-by #(get dists %) u))]
          (recur dists next-n (filter #(some (fn [s] (= s %)) u) (keys (get g next-n))) u p))
        (let [cdst (get dists curr) ; calculate new distance and update dists
              idist (get dists (first ns))
              sum (+ cdst (get (get g curr) (first ns)))
              result (if (< sum idist)
                       (assoc dists (first ns) sum)
                       dists)
              path (if (< sum idist) ; update path if shorter one is found
                     (assoc p (first ns) curr)
                     p)]
          (recur result curr (rest ns) unvisited path))))))

  "Reconstruct map to common representation:
  {A {B 10, C 20}, C {A 20}, B {A 10}}
  (map<city,map<city,price>>)"
(defn simplify-graph [graph]
  (let [g (get graph :forward)]
    (zipmap
     (keys g)
     (map #(zipmap
            (keys %)
            (map
             (fn [param1] (if
                            (> (deref (param1 :tickets)) 0)
                            (get param1 :price)
                            9999)) (vals %))) (vals g))))) ; if no tickets, price is set too high

;;implementation must be pure functional besides the transaction itself, tickets reference modification and
;;restarts monitoring (atom could be used for this)
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn book-tickets
  "Tries to book tickets and decrement appropriate references in route-map atomically
   returns map with either :price (for the whole route) and :path (a list of destination names) keys
          or with :error key that indicates that booking is impossible due to lack of tickets"
  [route-map from to]
  (if (= from to)
    {:path '(), :price 0} ; if already there, there is no need to travel
    (let [graph (simplify-graph route-map)
          path (dijkstra graph from to)]
      (try
        (dosync ; start transaction
         (loop [p (second path)] ; loop through the path
           (if (< (count p) 3)
             nil ; exit loop
             (let [r (:tickets (get (get (:forward route-map) (first p)) (second p)))]
               (do
                 (alter r dec) ; decrease tickets count, if there were no tickets, exception is thrown
                 (recur (next p))))))) ; move to the next flight
        {:path (second path), :price (first path)} ;
        (catch Exception e {:error e}))))) ; caught exception -> abort transaction and return error

;;;cities
(def spec1
  (-> empty-map
      (route "City1" "Capital" 200 5)
      (route "a" "Capital" 200 5)
      (route "Capital" "City1" 250 5)
      (route "City2" "Capital" 200 5)
      (route "Capital" "City2" 250 5)
      (route "City3" "Capital" 300 3)
      (route "Capital" "City3" 400 3)
      (route "City1" "Town1_X" 50 2)
      (route "Town1_X" "City1" 150 2)
      (route "Town1_X" "TownX_2" 50 2)
      (route "TownX_2" "Town1_X" 150 2)
      (route "Town1_X" "TownX_2" 50 2)
      (route "TownX_2" "City2" 50 3)
      (route "City2" "TownX_2" 150 3)
      (route "City2" "Town2_3" 50 2)
      (route "Town2_3" "City2" 150 2)
      (route "Town2_3" "City3" 50 3)
      (route "City3" "Town2_3" 150 2)))

(defn booking-future [route-map from to init-delay loop-delay]
  (future
   (Thread/sleep init-delay)
   (loop [bookings []]
     (Thread/sleep loop-delay)
     (let [booking (book-tickets route-map from to)]
       (if (booking :error)
         bookings
         (recur (conj bookings booking)))))))

(defn print-bookings [name ft]
  (println (str name ":") (count ft) "bookings")
  (doseq [booking ft]
    (println "price:" (booking :price) "path:" (booking :path))))

(defn run []
  ;;try to tune timeouts in order to all the customers gain at least one booking 
  (let [f1 (booking-future spec1 "City1" "City3" 0 1),
        f2 (booking-future spec1 "City1" "City2" 100 1),
        f3 (booking-future spec1 "City2" "City3" 100 1)]
    (print-bookings "City1->City3:" @f1)
    (print-bookings "City1->City2:" @f2)
    (print-bookings "City2->City3:" @f3)
    ))

(defn -main
  [& args]
  (run)
  (Thread/sleep 2000)
  (shutdown-agents))


src/nsu_labs/core.clj
(ns nsu-labs.core
    (:gen-class))

(use
 'clojure.test)

(defmacro dbg[x] `(let [x# ~x] (println "dbg:" '~x "=" x#) x#))

(defn parallel-filter
  ([predicate coll]
   (let [parts       (partition-all 2048 (take 16384 coll))
         black-magic (fn [coll2]
                       (doall
                        (map deref
                             (doall (map (fn [coll1] (future (doall (filter predicate coll1)))) coll2)))))]
     (lazy-seq
      (if (empty? coll)
        '()
        (concat (reduce concat (black-magic parts))
                (parallel-filter predicate (drop 16384 coll))))))))

(defn prime? [n]
  (if (even? n)
    false
    (let [root (num (int (Math/sqrt n)))]
      (loop [i 3]
        (if (> i root)
          true
          (if (zero? (mod n i))
            false
            (recur (+ i 2))))))))

(defn my-wierd-predicate [x]
  (and (prime? x) (= (mod x 1337) 22)))
(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Usual filter:")
  (println (time (doall (filter my-wierd-predicate (range 1000000)))))
  (println (time (doall (filter my-wierd-predicate (range 1000000)))))
  ;(println (time (doall (filter my-wierd-predicate (range 1000000)))))
  (println "My parallel filter:")
  (println (time (doall (parallel-filter my-wierd-predicate (range 1000000)))))
  (println (time (doall (parallel-filter my-wierd-predicate (range 1000000)))))
  ;(println (time (doall (parallel-filter my-wierd-predicate (range 1000000)))))
  (println "lazyness test")
  (println (take 1 (parallel-filter my-wierd-predicate (iterate inc 1))))
  (shutdown-agents))

